### ISSUE TITLE
UIDF but for types

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#5155]
A very common pattern I see in C++ binaries with heavy use of OOP is a dynamic cast or type check of some sort, followed by accessing offsets past the base class's size. To my knowledge, there is currently no good way to represent this in BN.

An example:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/0d0022a2-6f34-4c42-ab34-006cf573f9fa)

`i_1` is typed as `S1GameObject*`, the base class for `S1AfterImage`. The last line accesses a field on `S1AfterImage` after testing that the type is correct.

It would be nice if there was some way to inform BN about such type knowledge. I don't think it has to be a full-blown system where you can annotate a function as representing a dynamic type test, as this could quickly explode in complexity (e.g. this binary I'm working on has 4-5 different ways of doing type tests). I *think* it would be sufficient to just be able to tell BN that, if a given branch is taken, the type of a variable is changed.

**Comment by alexrp** [comment]
Presumably this was addressed by 56f778f?

**Comment by rssor** [comment]
Yes, this should be made possible by the forced variable versions feature. Right clicking on i_1 at some point should allow you to force a definition to be created, at which point you can give it a new name/type.

**Comment by psifertex** [comment]
If you're able to test and verify it solves the original goal, feel free to close the issue and let us know!

**Comment by alexrp** [comment]
Hmm, in the example in the OP, clicking "Force New Variable Version" on `i_1` doesn't *seem* to do anything, regardless of where I do it.

Also, just to note, I had to switch to HLIL to be able to get that option at all; seems like it hasn't been added to Pseudo C.

**Comment by rssor** [comment]
Interesting, for there to be _nothing_ happening is a failure mode I haven't seen yet, any chance you can share a repro?

MLIL might also be insightful as HLIL can do an enormous amount of folding.

**Comment by rssor** [comment]
Disregard, just spotted it.

**Comment by rssor** [comment]
Can you confirm if manually causing the function to be reanalyzed makes it appear? Making some sort of change to a var name or something should suffice.

**Comment by alexrp** [comment]
Triggering a reanalysis indeed makes it work!

![Image](https://github.com/user-attachments/assets/a4162ace-9f76-4c63-9848-d0edcf215ab3)

**Comment by rssor** [comment]
Pushed some fixes for the UI not causing updates and some other minor issues.

Thanks for confirming! Hoping this unblocks a class of usability problems when dealing with code like that.
